lib/auth/sessions.go

```diff
@@ -23,37 +23,32 @@ import (
 	"github.com/gravitational/teleport"
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/api/types/wrappers"
+	"github.com/gravitational/teleport/lib/backend"
 	"github.com/gravitational/teleport/lib/jwt"
 	"github.com/gravitational/teleport/lib/modules"
 	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local"
+	"github.com/gravitational/teleport/lib/tlsca"
 	"github.com/gravitational/teleport/lib/utils"
 
 	"github.com/gravitational/trace"
-
+	"github.com/jonboulle/clockwork"
 	"github.com/pborman/uuid"
 )
 
 // CreateAppSession creates and inserts a services.WebSession into the
 // backend with the identity of the caller used to generate the certificate.
 // The certificate is used for all access requests, which is where access
 // control is enforced.
-func (s *Server) CreateAppSession(ctx context.Context, req services.CreateAppSessionRequest, user services.User, checker services.AccessChecker) (services.WebSession, error) {
+func (s *Server) CreateAppSession(ctx context.Context, req services.CreateAppSessionRequest, user services.User, identity tlsca.Identity, checker services.AccessChecker) (services.WebSession, error) {
 	if !modules.GetModules().Features().App {
 		return nil, trace.AccessDenied(
 			"this Teleport cluster doesn't support application access, please contact the cluster administrator")
 	}
 
-	// Check that a matching parent web session exists in the backend.
-	parentSession, err := s.GetWebSession(ctx, types.GetWebSessionRequest{
-		User:      req.Username,
-		SessionID: req.ParentSession,
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	// Don't let the TTL of the child certificate go longer than the parent.
-	ttl := checker.AdjustSessionTTL(parentSession.GetExpiryTime().Sub(s.clock.Now()))
+	// Don't let the app session go longer than the identity expiration,
+	// which matches the parent web session TTL as well.
+	ttl := checker.AdjustSessionTTL(identity.Expires.Sub(s.clock.Now()))
```

Comment:
@r0mant I think this is fine.

Can you update the comment with some more details about why this works for browser based Application Access: browser will send a cookie with `sessionID` which will be used to fetch `services.WebSession` which contains a certificate whose life matches the life of the session that will be used to establish the connection.

@awly You might want to review these changes as well.

Reply:
Updated the comment.